apps/mux/src/config.tsx

```diff
@@ -0,0 +1,109 @@
+  }
+
+  async componentDidMount () {
+    // Get current parameters of the app.
+    const parameters = await this.app.getParameters();
+
+    this.setState(
+      // If the app is not installed, `parameters` will be `null`.
+      // We default to an empty object in this case.
+      { parameters: parameters || {} },
+      () => {
+        // Once preparation has finished, call `setReady` to hide
+        // the loading screen and present the app to a user.
+        this.app.setReady()
+      }
+    );
+  }
+
+  // Renders the UI of the app.
+  render () {
+    const { parameters: { muxAccessTokenId, muxAccessTokenSecret } } = this.state;
+
+    return (
+      <React.Fragment>
+        <div className='form-wrapper'>
+          <Form id="app-config" spacing="default">
+            <Heading>Mux Video Uploader</Heading>
+            <Note noteType="primary" title="About the app">
+              A Contentful UI app that makes it simple to add beautiful streaming via <a href='https://mux.com'>Mux</a> to your Contentful project. Install the extension, add the component to your content model, and you're good to go! 🙌
+            </Note>
+            <TextField
+              required
+              name="mux-access-token"
+              id="mux-access-token"
+              labelText="Mux access token"
+              value={muxAccessTokenId || ''}
+              onChange={e => this.setState({ parameters: { muxAccessTokenId: e.target.value, muxAccessTokenSecret } })}
+            />
+            <TextField
+              required
+              name="mux-token-secret"
+              id="mux-token-secret"
+              labelText="Mux token secret"
+              value={muxAccessTokenSecret || ''}
+              onChange={e => this.setState({ parameters: { muxAccessTokenId, muxAccessTokenSecret: e.target.value } })}
+            />
+          </Form>
+        </div>
+      </React.Fragment>
+    );
+  }
+
+  // This method will be called when a user clicks on "Install"
+  // or "Save" in the configuration screen.
+  //
+  // We want to do two things in here:
+  // 1. Persist selected animal as a parameter
+  // 2. Put the app to sidebars of all Content Types
+  async onConfigure () {
+    // Get IDs of all content types in an environment.
+    const { items: contentTypes } = await this.props.sdk.space.getContentTypes();
+    const contentTypeIds = contentTypes.map(ct => ct.sys.id)
+
+    // Return value of `onConfigure` is used to install
+    // or update the configuration.
+    return {
+      // Parameters to be persisted as the app configuration.
+      parameters: this.state.parameters,
+      // Transformation of an environment performed in the
+      // installation process.
+      targetState: {
+        EditorInterface: contentTypeIds.reduce((acc, id) => {
+          // Insert the app as the first item in sidebars
+          // of all content types.
+          return { ...acc, [id]: { sidebar: { position: 0 } } }
```

Comment:
This logic needs to change. Currently, you are adding Mux to the sidebar of every content type in the space. Mux, however, doesn't use the sidebar and will create a blank space for every single content type at the top.

Instead, you'll need to target the specific field inside of each content type where mux should show up as an `input` element. We do this in a few of our other apps.

For example, our other DAM apps share a library which helps contain this logic.

Where we set target state for DAM apps:
https://github.com/contentful/apps/blob/master/lib/shared-dam-app/src/AppConfig/AppConfig.tsx#L146

The utility functionality to help find target fields to attach to per content type:
https://github.com/contentful/apps/blob/master/lib/shared-dam-app/src/AppConfig/fields.ts

Reply:
> You'll need to target the specific field inside of each content type where mux should show up as an input element.

This isn't exactly clear to me - would you mind helping me understand please? Currently Mux appears as a "JSON object" field

![contentful-video-upload](https://user-images.githubusercontent.com/764988/81205599-c95aa180-8f7f-11ea-9be4-862a0c21bee1.gif)

Reply:
Looking at the code it appears that when you install the mux app, you'll still need the user to go into their content model and find the JSON fields where they want the mux UI to show up.  

If that is the case, we can create UI inside of the app configuration screen where the user can select these fields before installing the app. This means that a user won't have to go into the content model and manually apply mux to the JSON fields.

If you check out the Bynder app for example, we show the user a list of content types and fields within those content types where they can choose to install the Bynder app to. Our logic checks to ensure we only show JSON fields.

<img width="911" alt="Screen Shot 2020-05-07 at 12 53 41" src="https://user-images.githubusercontent.com/4086109/81286762-e071c000-9061-11ea-97b8-d06168054fca.png">

The UI above you will have to recreate yourself (you can check out the Bynder code) however, to save the user selection, we can apply this logic in the return value of your `onConfigure` handler that you pass in to the property `EditorInterface`.

```js

return {
 parameters: this.state.parameters,
 EditorInterface: {
    post: {controls: [{fieldId: 'blogImage'}], 
  }
}
```

The `EditorInterface` object's keys are the IDs of content types we want to configure for the app. In the code example above, `post` is the ID of a content type that has a JSON field with an ID of `blogImage`

Mux can use this same approach to enable users to get the mux app working for them right away after the installation without additional steps. We have a set of utility functions that we use for all our DAM apps that can help you shape these responses for the `EditorInterface` value.
 
They can be found here:

https://github.com/contentful/apps/blob/master/lib/shared-dam-app/src/AppConfig/fields.ts

I know this is a lot to cover and the docs are a bit light in this area. We are working to improve them but in the mean time please follow up with any questions and I can help you get on your way.

Reply:
@davidfateh Thanks! This all makes a lot of sense. I have this almost working - I created the "Assign to Fields" UI. I think I did this correctly - it only shows *content types* that have at least 1 JSON field, and under each content type it only shows the fields that are JSON fields.

<img width="859" alt="assign-to-fields_2020-05-07_09-58-49" src="https://user-images.githubusercontent.com/764988/81322834-60d7f700-9049-11ea-80f8-d0a51b9d6d7e.png">

I'm a little stuck now using the helpers to manage the local state and construct the `EditorInterface` value.

Bynder looks like a perfect example, but checking [the code in this repo](https://github.com/contentful/apps/blob/master/apps/bynder/src/index.js) - All I see is this 138 line index file -- which looks like the code for the UI. I can't find the config code, does it live somewhere else?

Reply:
Okay... I got closer here... but saving the `EditorInterface` value is not quite working. Here's an example where I save this

```
return {
  parameters,
  EditorInterface: {
    blogPost: {
      controls: [{ fieldId: "videoUpload" }]
    }
  }
}
```

I must be missing something really obvious here. `blogPost` is the content type ID and `videoUpload` is the field ID. This is a screenshot where I am logging the value that gets set for the `EditorInterface` key:

<img width="523" alt="editorInterface_2020-05-11_15-28-01" src="https://user-images.githubusercontent.com/764988/81618463-416c0180-939c-11ea-9b6d-8ae04aa196fb.png">

----

I saw in another example `EditorInterface` was [nested under the key `targetState`](https://github.com/contentful/apps/blob/master/apps/gatsby/src/AppConfig/AppConfig.js#L110) so I tried that too, but it didn't work either.

-----

Two outstanding things to get this wrapped up:

* What am I doing wrong above?
* I had to put in a lot of boilerplate code converting to and from the editorInterface/contentType data structures. How can I use your helper methods? What's confusing me is:
   1. [shared-dam-app](https://github.com/contentful/apps/blob/master/apps/bynder/package.json#L15) I see this package is using it, but it [doesn't exist on NPM](https://www.npmjs.com/package/shared-dam-app). Is it private or something?
  2. I see other apps using the `setup` method from `shared-dam-app`. Should I be using that?! It looks way easier for these configuration steps...

Reply:
Hey so I checked out your code and you are right: I missed `targetState` as the top level property on the return value. I also think the reason it wasn't working for you was due to older versions of our contentful packages.

In this PR I was able to get target state working and on refresh I can see my saved changes show up.
https://github.com/muxinc/apps/pull/1

As for your second question about the shared-dam-app packages, we use Lerna in this repo to manage local packages. So while they are not distributed via npm, Lerna will see that we have local shared packages and install them. The reason that Bynder, for example, only have UI in its config file is because of your observation about the shared libs.

I'm not sure if Mux will benefit from using it as your UX is different. The DAM apps pop a dialog modal which then iframes into their asset pickers. Do you have such an interface?

Reply:
@davidfateh fantastic! Good find!

okay, that all makes sense.